source/common/http/status.h

```diff
@@ -0,0 +1,148 @@
+ *   3. Register size, to avoid stack allocations.
+ *   4. Extensible.
+ *
+ * Constructing an error status is likely as heavy as throwing an exception due to a heap
+ * allocation.
+ */
+class ABSL_MUST_USE_RESULT Status final {
+public:
+  // Creates an OK status with no message.
+  Status() = default;
+
+  /**
+   * Create a status with the specified code and
+   * error message. If `code == StatusCode::Ok`, `msg` is ignored and an
+   * object identical to an OK status is constructed.
+   */
+  Status(StatusCode code, absl::string_view msg);
+
+  /**
+   * Create a status with the specified code, HTTP code and
+   * error message. If `code == StatusCode::Ok`, `msg` and `http_code` are ignored and an
+   * object identical to an OK status is constructed.
+   */
+  Status(StatusCode code, Http::Code http_code, absl::string_view msg);
+
+  Status(const Status& other) : rep_(other.rep_) { Ref(); }
+  Status& operator=(const Status& x);
+
+  // The moved-from state is the Ok.
+  Status(Status&& other) noexcept : rep_(other.rep_) { other.rep_ = nullptr; }
+  Status& operator=(Status&&) noexcept;
+
+  ~Status();
+
+  // Returns true if the Status is OK.
+  ABSL_MUST_USE_RESULT bool Ok() const { return rep_ == nullptr; }
+
+  // Returns the error code.
+  StatusCode Code() const { return Ok() ? StatusCode::Ok : rep_->status_code_; }
+
+  /**
+   * Returns the error message. Note: prefer ToString() for debug logging.
+   * This message rarely describes the error code. It is not unusual for the
+   * error message to be the empty string.
+   */
+  absl::string_view Message() const { return Ok() ? *EmptyString() : rep_->message_; }
+
+  // Returns HTTP code if it was provided during Status construction
+  absl::optional<Envoy::Http::Code> HttpCode() const {
+    return Ok() ? absl::nullopt : rep_->http_code_;
+  }
+
+  /**
+   * Returns a combination of the error code name, HTTP status if it was specified and the message.
+   * You can expect the code name and the message to be substrings of the
+   * result.
+   * WARNING: Do not depend on the exact format of the result of `ToString()`
+   * which is subject to change.
+   */
+  std::string ToString() const { return Ok() ? "OK" : ToStringError(); }
+
+  /**
+   * Ignores any errors. This method does nothing except potentially suppress
+   * complaints from any tools that are checking that errors are not dropped on
+   * the floor.
+   */
+  void IgnoreError() const {}
+
+private:
+  static const std::string* EmptyString();
+
+  // Returns string for non-ok Status.
+  std::string ToStringError() const;
+
+  struct StatusRep {
+    StatusRep(StatusCode code, absl::string_view msg)
+        : ref_(0), status_code_(code), message_(msg) {}
+    StatusRep(StatusCode code, Http::Code http_code, absl::string_view msg)
+        : ref_(0), status_code_(code), message_(msg), http_code_(http_code) {}
+
+    std::atomic<int32_t> ref_;
+    StatusCode status_code_;
```

Comment:
If we are going to go to the effort of importing this class due to lack of abseil extensibility, WDYT about making this a generic template class in which `status_code_` and `http_code_` could be put into an HTTP specific struct and then passed in that way? I think it would have basically the same perf and then we could use this elsewhere?

Also, should we briefly talk to the Abseil folks? Would they take a patch to do something similar which would avoid us having to maintain this code?

Reply:
done